[Deps] Update Kotlin to 2.0.21

ParaskP7 opened the pull request:
Project Thread: paaHJt-7n2-p2
Associated Metrics: paaHJt-7n2-p2#comment-9668

This PR updates Kotlin to 2.0 ([2.0.21](https://href.li/?https://github.com/JetBrains/kotlin/releases/tag/v2.0.21)).

## Description:

FYI: I recommend reviewing this PR commit-by-commit just to make sure that Kotlin has been updated appropriately and that there isn't any left-overs that need addressing.

PS: Main commits for some extra focus:
- [Build: Update/Add jvm target & source/target compatibility to java 11](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/6abe01ec983d0765d4bd1466030097109867ac56)
- [Analysis: Resolve val cannot be reassigned error on site store](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/7a9e63fef94728ac0046ff40ec2901f56e66c01b)
- [Analysis: Resolve non-public primary ctor warning on wordpress module](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/0ed041bf7b398c916fff41b22bb56156d9864627)
- [Deps: Update kotlin compile testing to 1.6.0 (tschuchortdev)](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/459eaa8a3d7cfc1214ba3641f45ba7bc4c82346e) (*)

(*) About `kotlin-compile-testing` ([tschuchortdev](https://github.com/tschuchortdev/kotlin-compile-testing)), it made me wonder how well maintained this library is and whether we can depend on in for future Kotlin updates, let's discuss. 🤔

-----

## To Test:

<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->

- Smoke test both, the `Jetpack` and `WordPress` apps, and verify everything is working as expected.
- Try to test for build time increases both, locally and on CI.
- You could also run the `./gradlew buildHealth` task locally and verify that everything is working as expected ([CI](https://buildkite.com/automattic/wordpress-android/builds/20336)).
- Review the [project dependencies & build environment changes](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#issuecomment-2460220867) and assert correctness.

-----

## Regression Notes

1. Potential unintended areas of impact

    - Build time increases (clean and/or incremental).

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist (strike-out the not-applying and unnecessary ones): N/A

ParaskP7 committed:
Deps: Update kotlin to 2.0.21 (ksp to 2.0.21-1.0.26)

Release Notes:
- Kotlin: https://github.com/JetBrains/kotlin/releases/tag/v2.0.21
- KSP: https://github.com/google/ksp/releases/tag/2.0.21-1.0.26

ParaskP7 committed:
Build: Move kotlin related source/target compatibility into kotlin block

This solves the 'Assignment is not used' build warnings on the
'libs:annotations' and 'libs:processors' Kotlin related modules.

ParaskP7 committed:
Build: Update/Add jvm target & source/target compatibility to java 11

This change solves the below KSP related compatibility failure with
version '2.0.21-1.0.26', which is only compatible with JVM runtime
version 11 or newer. Also change is also solving the inconsistent
JVM-target compatibility issues on a number of modules, just because it
was missing an explicit source/target compatibility build declaration,
those module are:
- libs:fluxc
- libs:login
- libs:editor
- libs:image-editor

For more info see failure below:
------------------------------------------------------------------------
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:processors:compileKotlin'.
> Could not resolve all files for configuration
   ':libs:processors:compileClasspath'.
   > Could not resolve com.google.devtools.ksp:symbol-processing-api:
      2.0.21-1.0.26.
     Required by:
         project :libs:processors
      > Dependency resolution is looking for a library compatible with
       JVM runtime version 8, but 'com.google.devtools.ksp:
       symbol-processing-api:2.0.21-1.0.26' is only compatible with JVM
       runtime version 11 or newer.
------------------------------------------------------------------------
FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':libs:editor:compileDebugKotlin'.
> Inconsistent JVM-target compatibility detected for tasks
   'compileDebugJavaWithJavac' (1.8) and 'compileDebugKotlin' (11).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation: https://kotl.in/gradle/jvm/target-validation

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':libs:image-editor:compileDebugKotlin'.
> Inconsistent JVM-target compatibility detected for tasks
   'compileDebugJavaWithJavac' (1.8) and 'compileDebugKotlin' (11).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation: https://kotl.in/gradle/jvm/target-validation
------------------------------------------------------------------------
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:fluxc:kaptGenerateStubsDebugKotlin'.
> Inconsistent JVM-target compatibility detected for tasks
   'compileDebugJavaWithJavac' (1.8) and 'kaptGenerateStubsDebugKotlin'
   (11).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation:
   https://kotl.in/gradle/jvm/target-validation
------------------------------------------------------------------------
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:login:kaptGenerateStubsDebugKotlin'.
> Inconsistent JVM-target compatibility detected for tasks
   'compileDebugJavaWithJavac' (1.8) and 'kaptGenerateStubsDebugKotlin'
   (11).

  Consider using JVM Toolchain: https://kotl.in/gradle/jvm/toolchain
  Learn more about JVM-target validation:
   https://kotl.in/gradle/jvm/target-validation

ParaskP7 committed:
Build: Migrate jvm target & src/target compatibility to version catalogs

With this change, changing the 'java' version value within the
'libs.versions.toml' next time, will be all it takes to update JVM to
point to '17' or beyond.

ParaskP7 committed:
Deps: Add compose compile gradle plugin

Starting in Kotlin 2.0, the Compose Compiler Gradle plugin is required
when compose is enabled. This change addresses that, along with removing
the oudated 'androidx-compose-compiler' configuration.

------------------------------------------------------------------------
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':WordPress'.
> com.android.builder.errors.EvalIssueException: Starting in Kotlin 2.0,
   the Compose Compiler Gradle plugin is required
  when compose is enabled. See the following link for more information:
  https://d.android.com/r/studio-ui/compose-compiler

ParaskP7 committed:
Analysis: Resolve val cannot be reassigned error on site store

Error Message: "'val' cannot be reassigned."

ParaskP7 committed:
Analysis: Suppress deprecated in java warnings on wordpress module

Warning Messages: "This declaration overrides a deprecated member but is
not marked as deprecated itself. Please add the '@Deprecated' annotation
or suppress the diagnostic."

ParaskP7 committed:
Analysis: Resolve non-public primary ctor warning on wordpress module

Warning Messages: "Non-public primary constructor is exposed via the
generated 'copy()' method of the 'data' class."

FYI: There is no good reason for this constructor to be 'internal',
that is, instead of it being the default 'public'.

ParaskP7 committed:
Analysis: Resolve then suppress safe call warnings on wordpress module

Warning Messages: "Only safe (?.) or non-null asserted (!!.) calls are
allowed on a nullable receiver of type 'kotlin.Array<(out)
CapturedType(in T!)!>?'."

FYI: This is done because unless putting the safe call on 'type', this
warning persist and can't be otherwise suppressed.

ParaskP7 committed:
Test: Fix 19 test failures on kotlin no when branch matched exception

------------------------------------------------------------------------

MigrationEmailHelperTest > Should call migrationComplete when
 notifyMigrationComplete is called FAILED
    kotlin.NoWhenBranchMatchedException at
     MigrationEmailHelperTest.kt:27

CampaignListingViewModelTest > given campaigns available, when clicked
 on campaign, then navigate to detail FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     CampaignListingViewModelTest.kt:146

ThreatDetailsViewModelTest > when ok button on fix action confirmation
 dialog is clicked, then action buttons are disabled FAILED
    kotlin.NoWhenBranchMatchedException at
     ThreatDetailsViewModelTest.kt:370

ThreatDetailsViewModelTest > when ok button on ignore action
 confirmation dialog is clicked, then action buttons are disabled FAILED
    kotlin.NoWhenBranchMatchedException at
     ThreatDetailsViewModelTest.kt:365

ReaderInterestsViewModelTest > selected interests saved on done button
 click FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderInterestsViewModelTest.kt:599

ReaderInterestsViewModelTest > get user tags re-triggered on retry if
 user tags request had failed earlier FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     ReaderInterestsViewModelTest.kt:625

ReaderInterestsViewModelTest > interest deselected if
 onInterestAtIndexToggled invoked on a selected interest FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     ReaderInterestsViewModelTest.kt:599

ReaderPostDetailViewModelTest > given local post not found, when show
 post is triggered, then loading state is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1117

ReaderPostDetailViewModelTest > given wp com post, when related posts
 are requested, then related posts are fetched FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:813

ReaderPostDetailViewModelTest > given request failed, when post is
 fetched, then request failed message is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given no network, when post is fetched,
 then no network message is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised with signin offer and
 no intercept uri, when error ui shown, then correct msg exists FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised with no signin offer
 and no intercept uri, when error ui shown, then correct msg exists
 FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given local post not found, when show
 post is triggered, then post is fetched from remote server FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1117

ReaderPostDetailViewModelTest > given unauthorised with no signin offer
 and intercept uri, when error ui shown, then correct msg exists FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised, when post is
 fetched, then error ui is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given request already running, when post
 is fetched, then no error is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised with signin offer,
 when error ui shown, then sign in button is visible FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised without signin
 offer, when error ui shown, then sign in button is not visible FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given post not found, when post is
 fetched, then post not found message is shown FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

ReaderPostDetailViewModelTest > given unauthorised with signin offer and
 intercept uri, when error ui shown, then correct msg exists FAILED
    kotlin.NoWhenBranchMatchedException at
     ReaderPostDetailViewModelTest.kt:1263

StatsViewModelTest > given enable stats module is clicked, then state
 reflects progress FAILED
    kotlin.NoWhenBranchMatchedException at
     StatsViewModelTest.kt:241

ParaskP7 committed:
Test: Fix 3 test failures on coroutines uncaught exceptions before test

------------------------------------------------------------------------

CampaignListingViewModelTest > given campaigns available, when clicked
 on campaign, then navigate to detail FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     CampaignListingViewModelTest.kt:146

ReaderInterestsViewModelTest > get user tags re-triggered on retry if
 user tags request had failed earlier FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     ReaderInterestsViewModelTest.kt:625

ReaderInterestsViewModelTest > interest deselected if
 onInterestAtIndexToggled invoked on a selected interest FAILED
    kotlinx.coroutines.test.UncaughtExceptionsBeforeTest at
     ReaderInterestsViewModelTest.kt:599

ParaskP7 committed:
Deps: Update kotlin compile testing to 1.6.0 (tschuchortdev)

Release Notes: https://github.com/tschuchortdev/kotlin-compile-testing/
releases/tag/1.6.0

This update was done mainly due to the below library
related incompatibility while testing 'RemoteConfigProcessorTest':

e: Incompatible classes were found in dependencies. Remove them from the
 classpath or use '-Xskip-metadata-version-check' to suppress errors
e: file:///.../Feature.kt:3:41 Class 'org.wordpress.android.annotation
 .Feature' was compiled with an incompatible version of Kotlin. The
 binary version of its metadata is 2.0.0, expected version is 1.8.0.
The class is loaded from /Users/.../WordPress-Android/libs/annotations/
 build/libs/annotations.jar!/org/wordpress/android/annotation/
 Feature.class

------------------------------------------------------------------------

Otherwise, this is done because 'RemoteConfigProcessorTest' was failing
with the below 1 x 'ClassNotFoundException' exception:

RemoteConfigProcessorTest > given class with feature in development
 annotation, when compiling, generate expected list of classes FAILED
    java.lang.ClassNotFoundException at RemoteConfigProcessorTest.kt:107

------------------------------------------------------------------------

While also, 'RemoteConfigProcessorTest' was failing with the below 3 x
'ComparisonFailure' test failures:

RemoteConfigProcessorTest > given a class with remote field annotation,
 when compiling, generate expected config defaults class FAILED
    org.junit.ComparisonFailure at RemoteConfigProcessorTest.kt:40

RemoteConfigProcessorTest > given a class with features annotation,
 when compiling, generate expected configuration check FAILED
    org.junit.ComparisonFailure at RemoteConfigProcessorTest.kt:18

RemoteConfigProcessorTest > given class with feature and experiment
 annotation, when compiling, generate config defaults class FAILED
    org.junit.ComparisonFailure at RemoteConfigProcessorTest.kt:70

------------------------------------------------------------------------

No matter, even after updating the library, the required
'ExperimentalCompilerApi' got added on 'RemoteConfigProcessorTest' to
fix a number of  'Experimental API' errors such as the below:

"This API is experimental. There are no stability guarantees for it"

As well as adding the 'UnsafeCastFunction' annotation to fix the below
3 x 'UNCHECKED_CAST'  errors:

------------------------------------------------------------------------

> Task :libs:processors:compileTestKotlin FAILED
e: file:///Users/.../WordPress-Android/libs/processors/src/test/kotlin/
 org/wordpress/android/processor/RemoteConfigProcessorTest.kt:53:18
        Usage of this function is unsafe because it does not have native
         compiler support
         This means that compiler won't report UNCHECKED_CAST,
          CAST_NEVER_SUCCEED or similar
         diagnostics in case of error cast (which can happen immediately
          or after some
         refactoring of class hierarchy)
        Consider using regular `as` and `as?`

e: file:///Users/.../WordPress-Android/libs/processors/src/test/kotlin/
 org/wordpress/android/processor/RemoteConfigProcessorTest.kt:83:18
        Usage of this function is unsafe because it does not have native
         compiler support
         This means that compiler won't report UNCHECKED_CAST,
          CAST_NEVER_SUCCEED or similar
         diagnostics in case of error cast (which can happen immediately
          or after some
         refactoring of class hierarchy)
        Consider using regular `as` and `as?`

e: file:///Users/.../WordPress-Android/libs/processors/src/test/kotlin/
 org/wordpress/android/processor/RemoteConfigProcessorTest.kt:117:18
        Usage of this function is unsafe because it does not have native
         compiler support
         This means that compiler won't report UNCHECKED_CAST,
          CAST_NEVER_SUCCEED or similar
         diagnostics in case of error cast (which can happen immediately
          or after some
         refactoring of class hierarchy)
        Consider using regular `as` and `as?`

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:processors:compileTestKotlin'.
> A failure occurred while executing org.jetbrains.kotlin.compilerRunner
 .GradleCompilerRunnerWithWorkers$GradleKotlinCompilerWorkAction
   > Compilation error. See log for more details

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 commented:
FYI: The [SonarCloud](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#issuecomment-2462217734) analysis above is failing only because of those [4 override functions](https://sonarcloud.io/project/issues?issueStatuses=OPEN%2CCONFIRMED&pullRequest=21428&types=CODE_SMELL&inNewCodePeriod=true&sinceLeakPeriod=true&id=wordpress-mobile_WordPress-Android) that got suppressed with the ` @Deprecated("Deprecated in Java")` annotation, should be too serious. 👍

nbradbury commented:
@ParaskP7 Thanks for the hard work on this! My first pass at testing didn't turn up any issues, but I do note SonarCloud flagged some deprecation issues. The initial build seemed a bit slow but subsequent builds were fine.

Regarding `kotlin-compile-testing`, I'm not at all familiar with that library but I'm always in favor of dropping dependencies that don't provide obvious value.

wzieba commented:
> (*) About kotlin-compile-testing ([tschuchortdev](https://github.com/tschuchortdev/kotlin-compile-testing)), it made me wonder how well maintained this library is and whether we can depend on in for future Kotlin updates, let's discuss. 🤔

> Regarding kotlin-compile-testing, I'm not at all familiar with that library but I'm always in favor of dropping dependencies that don't provide obvious value.

We use this library to unit test processor code generation logic from `processors` module. I think this dependency doesn't look like not maintained but if there's a reason you'd like to drop it, maybe we can migrate to this fork: https://github.com/zacsweers/kotlin-compile-testing 

> The goal of this fork is to better track the latest Kotlin releases.

ParaskP7 commented:
👋 @nbradbury !

> @ParaskP7 Thanks for the hard work on this! My first pass at testing didn't turn up any issues...

Awesome, thanks for taking a look at it and going through testing! 🙇 🧪 🚀 

> ...but I do note SonarCloud flagged some deprecation issues.

Yea, as per [my comment](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#issuecomment-2462519264) above I think we can safely ignore that.

>  The initial build seemed a bit slow but subsequent builds were fine.

Great to hear! 👍 

> Regarding kotlin-compile-testing, I'm not at all familiar with that library but I'm always in favor of dropping dependencies that don't provide obvious value.

Actually, this dependency does provide (obvious) value, it helps us test our annotation processors (see [RemoteConfigProcessorTest.kt](https://github.com/wordpress-mobile/WordPress-Android/blob/73b63d69704ec570d9a7777918e5c8de92969a38/libs/processors/src/test/kotlin/org/wordpress/android/processor/RemoteConfigProcessorTest.kt)). My concern about this library is whether it is well maintained and what will happen on subsequent Kotlin/KSP updates if that's not the case. 🤔

ParaskP7 committed:
Deps: Update ksp to 2.0.21-1.0.27

Release Notes: https://github.com/google/ksp/releases/tag/2.0.21-1.0.27

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into deps/update-kotlin-to-2.0.21

ParaskP7 commented:
FYI: @nbradbury @wzieba I just notice that yesterday the latest [2.0.21-1.0.27](https://github.com/google/ksp/releases/tag/2.0.21-1.0.27) KSP release was tagged. It includes some bugs fixes only. I didn't want us to miss this opportunity to have that here too, thinking it is safer to go with more bug fixes on such a library than not. Thus, I just pushed a new commit with this change:

- [Deps: Update ksp to 2.0.21-1.0.27](https://github.com/wordpress-mobile/WordPress-Android/pull/21428/commits/5d46d0bfcca9fdb5712669e8b7220e43f5b07d4e)

ParaskP7 commented:
👋 @wzieba and [jinx](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#issuecomment-2464256629)!

> We use this library to unit test processor code generation logic from processors module.

💯 

> I think this dependency doesn't look like not maintained...

What makes you think that this dependency is maintained, just to understand more about it? 🤔 

FYI: I only reasoned that it is not well maintained because the latest release ([1.6.0](https://github.com/tschuchortdev/kotlin-compile-testing/releases/tag/1.6.0)) was on May 26 (almost half a year ago), it updated Kotlin to 1.9.24 and not accounting for Kotlin 2.0 (at least it seems so). 🤷 

> ...but if there's a reason you'd like to drop it, maybe we can migrate to this fork: https://github.com/zacsweers/kotlin-compile-testing

TIL about Zac Sweers's version of [kotlin-compile-testing](https://github.com/ZacSweers/kotlin-compile-testing). This library seems indeed a bit more up-to-date, with its [0.5.0](https://github.com/ZacSweers/kotlin-compile-testing/releases/tag/0.5.0) release updating Kotlin to 2.0. 🤷

wzieba commented:
> What makes you think that this dependency is maintained, just to understand more about it? 🤔
FYI: I only reasoned that it is not well maintained because the latest release ([1.6.0](https://github.com/tschuchortdev/kotlin-compile-testing/releases/tag/1.6.0)) was on May 26 (almost half a year ago), it updated Kotlin to 1.9.24 and not accounting for Kotlin 2.0 (at least it seems so). 🤷

I wouldn't cosnider a dependency that had a release half a year ago not maintained. It's not marked archived or deprecated, it works well and brings value to the project. As it has complete set of features, at least for our case, I don't see why they should release more often. But anyway, I think for this specific reason, the linked above fork is maintained, so I'm completely fine if you'd like us to migrate to it instead!

wzieba commented on WordPress/src/test/java/org/wordpress/android/ui/stats/refresh/StatsViewModelTest.kt:
```diff
@@ -240,11 +240,12 @@ class StatsViewModelTest : BaseUnitTest() {
     @Test
     fun `given enable stats module is clicked, then state reflects progress`() = test {
         val uiModel = initObservers().uiModelObserver
+        whenever(statsModuleActivateUseCase.postActivateStatsModule(anyOrNull())).thenReturn(success)
 
         startViewModel(statsModuleEnabled = false)
         viewModel.onEnableStatsModuleClick()
 
-        assertThat(uiModel.last().disabledStatsProgressVisible).isTrue
+        assertThat(uiModel.last().disabledStatsProgressVisible).isFalse()
     }
```
The test after change tests something different than initially. I think we shouldn't change te test body, but instead fix the issue that breaks this test which is `kotlin.NoWhenBranchMatchedException` in `StatsViewModel#onEnableStatsModuleClick` method.

ParaskP7 replied:
Yes, this is exactly my thought process @wzieba , and I did briefly tried to make it (somehow) work, timeboxing my efforts, but in this specific situation, the test was wrong, the `last` value of `disabledStatsProgressVisible` ends up being `false` by the end of the test. As such, the `true` value is only set momentarily. This test was only working as is because the `when` branch wasn't working as expected. Does that make sense? 🤔

wzieba replied:
> As such, the true value is only set momentarily

That's true, and I think this is what the test wanted to test: that after `onEnableStatsModuleClick` is tapped, and `postActivateStatsModule` method is not finished, we should show "disabled stats progress". 

If you add an empty `else` to `onEnableStatsModuleClick` to fix `NoWhenBranchMatchedException` and revert changes to this test made in this PR, the test will pass. 

I prefer we do this, because `NoWhenBranchMatchedException` is a runtime exception, so we would start throwing it in production app if not fixed.

wzieba replied:
Ah I see, mockito returns null for `postActivateStatsModule` and this seems to trigger `NoWhenBranchMatchedException`, so not really a production-possible scenario

ParaskP7 replied:
😃 That's there is the problem @wzieba , adding an `else` there is wrong, it being redundant and all, this `when` is already covering all cases of `StatsModuleActivateRequestState`, which are:
- Success
- Failure.NetworkUnavailable
- Failure.RemoteRequestFailure

The problem with the test was that they didn't specify the `whenever(...).thenReturn(...)` test setup (`given`) to address the path, thus the test was failing on `NoWhenBranchMatchedException`. Now, no matter the path you chose, in the end the `disabledStatsProgressVisible` end up being `false`.

Does that help clarify? 🙏

wzieba replied:
Either way, to not alter test logic and test that `disabledStatsProgressVisible` is indeed `true` without going to `when` (which is now problematic) I'd suggest using the `StandardTestDispatcher` instead:

```kotlin
   @Test
    fun `given enable stats module is clicked, then state reflects progress`() = runTest(StandardTestDispatcher()) {
        val uiModel = initObservers().uiModelObserver

        startViewModel(statsModuleEnabled = false)
        viewModel.onEnableStatsModuleClick()

        assertThat(uiModel.last().disabledStatsProgressVisible).isTrue()
    }
```

ParaskP7 replied:
Why do you think this solves the problem? 🤔

1. I tried it just now and I still get the `kotlin.NoWhenBranchMatchedException` thrown.
2. IMHO we are having a wrong setup, not showing a clear path for the `when` case to progress further, thus an incomplete test which are hard to reason about.

wzieba replied:
Right, I forgot to remove `else` I've added previously when testing this change :).

I thought it will not run the coroutine declared in `launch` at line 392, so `NoWhenBranchMatchedException` won't be triggered.

wzieba replied:
No other ideas from me, this should be probably written differently to have a robust test for checking if "disabled state progress" has been shown. Thanks for the discussion, closing as resolved!

ParaskP7 replied:
Exactly my thinking, thus, I first considered removing it altogether, but then I chose to update it. Maybe we should actually remove it, just to keep things simple, wdyt @wzieba ? 🤔

wzieba commented:
Let's add `.kotlin` directory to `.gitignore` in scope of this PR - it's generated during build time.

ParaskP7 commented:
> Let's add .kotlin directory to .gitignore in scope of this PR - it's generated during build time.

Oh, true, good catch, I somehow missed that! 💯

ParaskP7 committed:
Git: Add .kotlin directory to .gitignore

This is done in order to avoid having the generated during build time
Kotlin session files being seen by Git.

PR Comment: https://github.com/wordpress-mobile/WordPress-Android/pull/
21428#issuecomment-2464658902

wzieba reviewed (approved):
Great work @ParaskP7 👏 ! The PR LGTM. Frankly, I disslike the neccesity of updating tests logic when updating Kotlin version, but I understand it was necessary 👍.

ParaskP7 commented:
> Great work @ParaskP7 👏 ! The PR LGTM. 

Thanks @wzieba !

> Frankly, I disslike the neccesity of updating tests logic when updating Kotlin version, but I understand it was necessary 👍.

`disslike` with double `s` really pin-points how much you dislike it, same here, it just that when working with mocking frameworks everything becomes that much more complicated. For that [specific test](https://github.com/wordpress-mobile/WordPress-Android/pull/21428#discussion_r1834170000) that we talked about, it should have been probably written in a better way, maybe we should actually delete it... 🤷

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into deps/update-kotlin-to-2.0.21

ParaskP7 commented:
👋 @nbradbury !

> @ParaskP7 Thanks for the hard work on this! My first pass at testing didn't turn up any issues...

Would you like to go for another pass at testing this or should I proceed with merging this to `trunk`? 🤔

nbradbury commented:
> Would you like to go for another pass at testing this or should I proceed with merging this to `trunk`?

I'm good with this being merged.

ParaskP7 commented:
Awesome, thanks for the confirmation Nick, merged! 🚀
